### DISCUSSION TITLE
Libmetal metal_device_open or metal_device_io_region error.

### DISCUSSION THREAD

**Discussion by Max4well** [OpenAMP/libmetal#300] [Q&A]
I have a petalinux system and a few apps running at startup. Sometimes one or more applications get a libmetal error. I received errors from these functions "[_metal_device_open_](https://github.com/OpenAMP/libmetal/blob/3aee6be866b190d2e2b4997fedbd976a0c37c0c6/lib/device.h#L130)" or"[_metal_device_io_region_](https://github.com/OpenAMP/libmetal/blob/3aee6be866b190d2e2b4997fedbd976a0c37c0c6/lib/device.h#L147)".  It happens once in 10 or 20 starts. When I received that error I tried to rerun the application manually and it ran fine. I only use the libmetal library to create shared memory between applications.

What could be the reason why I get this error at startup and not when I run it manually?
What is the reason why it happens occasionally?

**Comment by tnmysh** [comment]
That's interesting. Could you post what error messages you get?

**Comment by Max4well** [comment]
I received error in my _initlibmetal_ function.

`[#define](javascript:void(0);) SHM_DEV_NAME    "3ed80000.shm"

[#define](javascript:void(0);) BUS_NAME        "platform"

 

int initlibmetal(void){

    int ret;

 

    struct metal_device *device = NULL;

 

    ret = sys_init();

    if (ret) {

        LPERROR("Failed to initialize libmetal\n");

        return ret;

    }

 

    ret = 0;

 

    /* Open the shared memory device */

    ret = metal_device_open(BUS_NAME, SHM_DEV_NAME, &device);

    if (ret) {

        LPERROR("Failed to open device %s.\n", SHM_DEV_NAME);

        return ret;

    }

 

    /* get shared memory device IO region */

    io = metal_device_io_region(device, 0);

    if (!io) {

        LPERROR("Failed to get io region for %s.\n", device->name);

        ret = -ENODEV;

        if (device)

            metal_device_close(device);

    }

 

    return ret;

}

 

 

int sys_init()

{

    struct metal_init_params init_param = METAL_INIT_DEFAULTS;

    int ret;

 

    ret = metal_init(&init_param);

    if (ret)

        LPERROR("Failed to initialize libmetal\n");

    return ret;

}`

I either get "_Failed to open device 3ed80000.shm_" or "_Failed to get io region for 3ed80000.shm_" .
